packages/utilities/src/classNamesFunction.ts

```diff
@@ -1,33 +1,138 @@
+
+// Note that because of the caching nature within the classNames memoization,
+// I've disabled this rule to simply be able to work with any types.
+// tslint:disable:no-any
+
+// This represents a prop we attach to each Map to indicate the cached return value
+// associated with the graph node.
+const RetVal = '__retval__';
+
+interface IRecursiveMemoNode extends Map<any, IRecursiveMemoNode> {
+  [RetVal]?: string;
+}
+
+export interface IClassNamesFunctionOptions {
+  /**
+   * Disables class caching for scenarios where styleProp parts mutate frequently.
+   */
+  disableCaching?: boolean;
+}
+
 /**
  * Creates a getClassNames function which calls getStyles given the props, and injects them
  * into mergeStyleSets.
+ *
+ * Note that the props you pass in on every render should be in the same order and
+ * immutable (numbers, strings, and booleans). This will allow the results to be memoized. Violating
+ * these will cause extra recalcs to occur.
  */
-export function classNamesFunction<TStyleProps extends {}, TStyleSet extends IStyleSet<TStyleSet>>(): (
-  getStyles: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
-  styleProps?: TStyleProps
-) => IProcessedStyleSet<TStyleSet> {
-  // TODO: memoize.
+export function classNamesFunction<TStyleProps extends {}, TStyleSet extends IStyleSet<TStyleSet>>(
+  options: IClassNamesFunctionOptions = {}
+): (getStyles: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined, styleProps?: TStyleProps) => IProcessedStyleSet<TStyleSet> {
+  // We build a trie where each node is a Map. The map entry key represents an argument
+  // value, and the entry value is another node (Map). Each node has a `__retval__`
+  // property which is used to hold the cached response.
+
+  // To derive the response, we can simply ensure the arguments are added or already
+  // exist in the trie. At the last node, if there is a `__retval__` we return that. Otherwise
+  // we call the `getStyles` api to evaluate, cache on the property, and return that.
+  let map: IRecursiveMemoNode = new Map();
+  let resultCount = 0;
 
   const getClassNames = (
     styleFunctionOrObject: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
     styleProps: TStyleProps = {} as TStyleProps
   ): IProcessedStyleSet<TStyleSet> => {
-    // styleSet might be undefined if styleFunctionOrObject is undefined, but getStyles should never
-    // ordinarily be undefined (it would hardly make any sense).
-    // However, because we usually use `props.styles` as the argument to an invocation of this method, and
-    // `props.styles` itself is defined as optional, this avoids the need to use `!` at all invocation points.
-    if (styleFunctionOrObject === undefined) {
-      return {} as IProcessedStyleSet<TStyleSet>;
+    let current: Map<any, any> = map;
+    const disableCaching = options.disableCaching;
+
+    if (!options.disableCaching) {
+      current = _traverseMap(map, styleFunctionOrObject as any);
```

Comment:
What happens here if `styleFunctionOrObject` is a function? It seems like `_traverseMap` will traverse keys of the function, but is that right? Shouldn't it traverse the style function output?

Reply:
Ahh, I think I see now. You can use the function as a key since you're evaluating it against the `styleProps` below. If either change, it will traverse the map appropriately.